File: mobile/android/app/src/main/java/com/buonolo/app/data/ws/WsGateway.kt
package com.buonolo.app.data.ws

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.util.concurrent.TimeUnit
import kotlin.math.min

data class WsEvent(val type: String, val payload: JsonElement)

/** Thin wrapper around the backend's plain-WebSocket gateway with exponential-backoff reconnect. */
class WsGateway(private val wsBaseUrl: String) {
    private val client = OkHttpClient.Builder().pingInterval(30, TimeUnit.SECONDS).build()
    private val json = Json { ignoreUnknownKeys = true }
    private var socket: WebSocket? = null
    private var retryDelayMs = 1000L
    private var stopped = false
    private var currentToken: String? = null

    private val _events = MutableSharedFlow<WsEvent>(extraBufferCapacity = 64)
    val events: SharedFlow<WsEvent> = _events

    fun connect(accessToken: String) {
        stopped = false
        currentToken = accessToken
        openSocket(accessToken)
    }

    private fun openSocket(accessToken: String) {
        val request = Request.Builder().url("$wsBaseUrl/ws?token=$accessToken").build()
        socket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: okhttp3.Response) {
                retryDelayMs = 1000
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                runCatching {
                    val obj = json.parseToJsonElement(text).jsonObject
                    val type = obj["type"]?.jsonPrimitive?.content ?: return
                    val payload = obj["payload"] ?: return
                    _events.tryEmit(WsEvent(type, payload))
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                scheduleReconnect()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: okhttp3.Response?) {
                scheduleReconnect()
            }
        })
    }

    private fun scheduleReconnect() {
        if (stopped) return
        val token = currentToken ?: return
        Thread {
            Thread.sleep(retryDelayMs)
            retryDelayMs = min(retryDelayMs * 2, 30_000)
            if (!stopped) openSocket(token)
        }.start()
    }

    fun disconnect() {
        stopped = true
        socket?.close(1000, "client closing")
        socket = null
    }
}

File: mobile/ios/Buonolo/Networking/WsGateway.swift
import Foundation

struct WsEvent {
    let type: String
    let payload: Data
}

/// Thin wrapper around the backend's plain-WebSocket gateway with exponential-backoff reconnect.
actor WsGateway {
    static let shared = WsGateway()
    private init() {}

    private var task: URLSessionWebSocketTask?
    private var retryDelaySeconds: UInt64 = 1
    private var stopped = false
    private var currentToken: String?
    private var continuation: AsyncStream<WsEvent>.Continuation?

    lazy var events: AsyncStream<WsEvent> = AsyncStream { continuation in
        self.continuation = continuation
    }

    func connect(accessToken: String) async {
        stopped = false
        currentToken = accessToken
        await openSocket(accessToken: accessToken)
    }

    private func openSocket(accessToken: String) async {
        var components = URLComponents(url: await APIClient.shared.wsBaseURL, resolvingAgainstBaseURL: false)!
        components.path = "/ws"
        components.queryItems = [URLQueryItem(name: "token", value: accessToken)]
        guard let url = components.url else { return }

        let newTask = URLSession.shared.webSocketTask(with: url)
        task = newTask
        newTask.resume()
        retryDelaySeconds = 1
        listen(on: newTask)
    }

    private func listen(on task: URLSessionWebSocketTask) {
        task.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                Task { await self.handle(message) }
                Task { await self.listen(on: task) }
            case .failure:
                Task { await self.scheduleReconnect() }
            }
        }
    }

    private func handle(_ message: URLSessionWebSocketTask.Message) {
        guard case .string(let text) = message, let data = text.data(using: .utf8) else { return }
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = object["type"] as? String,
              let payloadObject = object["payload"] else { return }
        guard let payloadData = try? JSONSerialization.data(withJSONObject: payloadObject) else { return }
        continuation?.yield(WsEvent(type: type, payload: payloadData))
    }

    private func scheduleReconnect() async {
        guard !stopped, let token = currentToken else { return }
        try? await Task.sleep(nanoseconds: retryDelaySeconds * 1_000_000_000)
        retryDelaySeconds = min(retryDelaySeconds * 2, 30)
        if !stopped { await openSocket(accessToken: token) }
    }

    func disconnect() {
        stopped = true
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }
}
